### ISSUE TITLE
UndefinedBehaviorSanitizer incorrectly flags __builtin_clz(0) as undefined behavior

### ISSUE THREAD

**Issue by aaronpuchert** [llvm/llvm-project#66546]
This is easy to reproduce:
```c
int main(int argc, char** argv) {
  int value = __builtin_clz(0);
  return 0;
}
```
Now compile `-fsanitize=undefined` and run:
```
clz.c:2:29: runtime error: passing zero to clz(), which is not a valid argument
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior clz.c:2:29 in
```
But I don't think this is correct. The [GCC documentation says](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fclz):

>  If _x_ is 0, the result is undefined.

That's not undefined behavior, just an undefined result. The [LLVM documentation for the IR intrinsic](https://llvm.org/docs/LangRef.html#llvm-ctlz-intrinsic) states it more clearly:

> If `src == 0` then the result is the size in bits of the type of `src` if `is_zero_poison == 0` and `poison` otherwise.

Which is saying the same thing. So `__builtin_clz(0)` is not by itself undefined behavior.

I don't think it's necessarily wrong that UBSan warns about this, but it shouldn't be under `undefined-behavior`.

**Comment by shafik** [comment]
So the documentation for UBsan says: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

> -fsanitize=builtin: Passing invalid values to compiler builtins.

and if we use `-fno-sanitize=builtin` it stops complaining: https://godbolt.org/z/hdj5d48Ke

so it looks like it is behaving as intended.

**Comment by vitalybuka** [comment]
> If `src == 0` then the result is the size in bits of the type of `src` if `is_zero_poison == 0` and `poison` otherwise.

-fsanitize=undefined is about C/C++ UB, not about LLVM UB.

**Comment by aaronpuchert** [comment]
> So the documentation for UBsan says: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
> 
> > -fsanitize=builtin: Passing invalid values to compiler builtins.

That's also what the message says, but what does “invalid” mean? This case isn't undefined behavior, you just have to be careful about using the result.

> -fsanitize=undefined is about C/C++ UB, not about LLVM UB.

True, but those builtins are not part of (ISO) C/C++, and I've quoted their GCC documentation. That's probably as close to source semantics as we'll get, unless someone can find something better. (I can find no documentation on the Clang side, and the C++20 functions are a different matter, as they seem to have defined results.)

If the undefined behavior sanitizer wants to have no false positives, I think it cannot warn on the example. This is not UB.

**Comment by nikic** [comment]
I think you're reading too much into a quirk of the wording here. We read "the result is undefined" as "the behavior is undefined" here -- this documentation would need to be much more extensive to explain what exactly they mean by an "undefined result" and how it behaves operationally to be the intended reading, because I don't believe this notion is defined by the C/C++ standard.

The GCC documentation should probably be clarified to replace "result" with "behavior" to get this on firm normative ground.

**Comment by aaronpuchert** [comment]
> I think you're reading too much into a quirk of the wording here.

Hmm, that's possible. I just found it odd that it doesn't use the key phrase "undefined behavior", and that it also matches with the behavior of our middle-end intrinsic.

> I don't believe this notion is defined by the C/C++ standard.

True, but since builtins are language extensions, they might as well incorporate additional semantics on top of the standard. Though I guess to notion of `poison` is meant to remain internal to the middle end and not be exposed in the frontend. If only for the possibility of changing the semantics, as my impression is that we [haven't completely figured those out](http://www.cs.utah.edu/~regehr/papers/undef-pldi17.pdf) yet.
